unsloth/models/llama.py

```diff
@@ -518,6 +518,13 @@ def LlamaAttention_fast_forward(
         V = V.transpose(1, 2)
         A = flash_attn_func(Q, K, V, causal = True)
     else:
+        # when qlen==vlen and attn_mask is None, we should use causal attention
+        Q_len = Q.shape[-2]
+        K_len = K.shape[-2]
+        if attention_mask is None and Q_len == K_len:
+            is_causal = True
+        else:
+            is_causal = False
```

Comment:
May also need to change the usage of the `is_causal` argument in `scaled_dot_product_attention` function, such as here: https://github.com/unslothai/unsloth/pull/2868/files#diff-a45b72bb533eda979990bd79cde5fe9c9fde424779a4f1fc1195b75853d93b45R532